### PR TITLE
[symm_mem] Validate peer rank in put_signal/wait_signal, complete channel OOB tests

### LINKED ISSUE #191618: [symm_mem] check_channel() multiplies by world_size instead of dividing, allowing out-of-bounds signal writes which silently corrupts peer tensor data on main
### 🐛 Describe the bug

`check_channel()` computes the channel capacity as:

https://github.com/pytorch/pytorch/blob/050ce9e1226b95d6b5938bdd5e245491f7219fd8/torch/csrc/distributed/c10d/symm_mem/CUDASymmetricMemory-inl.cuh#L115-L131

```cpp
const size_t num_channels =
    signal_pad_size / sizeof(uint32_t) * world_size;
```

The signal pad holds `signal_pad_size / sizeof(uint32_t)` slots per rank, and each channel consumes `world_size` slots (slot index = `world_size * channel + rank`, see `barrier_kernel` just below in the same file), so the capacity should be:

```cpp
const size_t num_channels =
    signal_pad_size / sizeof(uint32_t) / world_size;
```

With the default 9216-byte pad and world_size = 2, the real capacity is 1152 channels, but the check accepts channels up to 4607 -- a factor of world_size^2 too permissive. `check_channel` is shared by the CUDA and NCCL backends, so `barrier()`, `put_signal()` and `wait_signal()` on both accept out-of-range channels.

## Consequence

Since #189088 (https://github.com/pytorch/pytorch/commit/c99d06a01764f62f60e3d1246fb026e63146e026) the signal pad sits at the front of the symmetric allocation, directly before the data buffer, so a signal op on an over-capacity channel writes its `1` past the pad straight into the peer's tensor data, silently.

Verified on 2x H200 (NVLink), main @ 050ce9e1226:

```
pad=9216B slots=2304 real_capacity=1152 channel=2000 -> writes 6784B past the pad
put_signal(channel=2000) was ACCEPTED (check_channel passed)
[rank 1] DATA CORRUPTED: t[1696] bits=0x00000001 (predicted victim element 1696)
```

The corrupted element is exactly where the slot arithmetic predicts. A correct bounds check would have raised instead.

## Repro

```python
# torchrun --standalone --nproc-per-node=2 repro.py
import os, torch, torch.distributed as dist
import torch.distributed._symmetric_memory as symm_mem

rank = int(os.environ["RANK"])
torch.cuda.set_device(rank)
dist.init_process_group("nccl")

t = symm_mem.empty(4096, dtype=torch.float32, device=f"cuda:{rank}")
t.zero_()
hdl = symm_mem.rendezvous(t, dist.group.WORLD)
dist.barrier()

if rank == 0:
    # capacity with default pad and ws=2 is 1152; 2000 passes the broken check
    hdl.put_signal(dst_rank=1, channel=2000)
torch.cuda.synchronize()
dist.barrier()

if rank == 1:
    print("nonzero elements:", t.nonzero().flatten())  # expected: none
```

## History / affected versions

The incorrect bound is unchanged since the original SymmetricMemory prototype (#128582, June 2024):

https://github.com/pytorch/pytorch/blob/8771e3429c3d7327f08c48d547ad73546d5603b3/torch/csrc/distributed/c10d/CUDASymmetricMemory.cu#L214

so every release with symmetric memory accepts out-of-range channels. The *consequence* differs by layout:

- Released versions use the old `[data buffer | signal pad]` layout, where the overflow lands past the end of the allocation -- in the granularity-rounded tail padding for moderate channels, or past the mapped VA for large ones.
- On main, after #189088 moved the pad to the front (`[signal pad | data buffer]`), the overflow lands in the peer's tensor data. Unless fixed, 2.14 will be the first release where an over-capacity channel silently corrupts user data.

## Fix

I can send a PR with the fix plus a test asserting that an over-capacity channel raises on both the CUDA and NCCL backends.

### Versions

```
PyTorch version: 2.14.0a0+gitdd30fd7
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: CentOS Stream 9 (x86_64)
GCC version: (GCC) 11.5.0 20240719 (Red Hat 11.5.0-11)
Clang version: Could not collect
CMake version: version 4.3.2
Libc version: glibc-2.34

Python version: 3.14.4 | packaged by Anaconda, Inc. | (main, Apr 14 2026, 17:07:44) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-5.14.0-615.el9.x86_64-x86_64-with-glibc2.34
Is CUDA available: True
CUDA runtime version: 13.0.88
CUDA_MODULE_LOADING set to: 
GPU models and configuration: 
GPU 0: NVIDIA H200
GPU 1: NVIDIA H200
GPU 2: NVIDIA H200
GPU 3: NVIDIA H200
GPU 4: NVIDIA H200
GPU 5: NVIDIA H200
GPU 6: NVIDIA H200
GPU 7: NVIDIA H200

Nvidia driver version: 580.82.07
cuDNN version: Probably one of the following:
/usr/lib64/libcudnn.so.9.13.0
/usr/lib64/libcudnn_adv.so.9.13.0
/usr/lib64/libcudnn_cnn.so.9.13.0
/usr/lib64/libcudnn_engines_precompiled.so.9.13.0
/usr/lib64/libcudnn_engines_runtime_compiled.so.9.13.0
/usr/lib64/libcudnn_graph.so.9.13.0
/usr/lib64/libcudnn_heuristic.so.9.13.0
/usr/lib64/libcudnn_ops.so.9.13.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  160
On-line CPU(s) list:                     0-159
Vendor ID:                               GenuineIntel
Model name:                              Intel Xeon Processor (SapphireRapids)
CPU family:                              6
Model:                                   143
Thread(s) per core:                      2
Core(s) per socket:                      40
Socket(s):                               2
Stepping:                                4
BogoMIPS:                                4200.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx_vnni avx512_bf16 wbnoinvd arat vnmi avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b fsrm md_clear serialize tsxldtrk amx_bf16 avx512_fp16 amx_tile amx_int8 arch_capabilities
Virtualization:                          VT-x
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               5 MiB (160 instances)
L1i cache:                               5 MiB (160 instances)
L2 cache:                                320 MiB (80 instances)
L3 cache:                                32 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-79
NUMA node1 CPU(s):                       80-159
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Mitigation; Aligned branch/return thunks
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Unknown: No mitigations
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                     Not affected
Vulnerability Tsx async abort:           Not affected

Versions of relevant libraries:
[pip3] nccl4py==0.3.1
[pip3] numpy==2.4.4
[pip3] optree==0.19.0
[pip3] torch==2.14.0a0+gitdd30fd7
[pip3] triton==3.6.0
[conda] cuda-cudart               13.0.96              ha533d76_1    nvidia
[conda] cuda-cudart-dev           13.0.96              h3b4bcfc_1    nvidia
[conda] cuda-cudart-dev_linux-64  13.0.96              hbe36340_1    nvidia
[conda] cuda-cudart-static        13.0.96              h3b4bcfc_1    nvidia
[conda] cuda-cudart-static_linux-64 13.0.96              hbe36340_1    nvidia
[conda] cuda-cudart_linux-64      13.0.96              h73695a4_1    nvidia
[conda] cuda-cupti                13.0.85              hf384b4a_0    nvidia
[conda] cuda-cupti-dev            13.0.85              hf384b4a_0    nvidia
[conda] cuda-libraries            13.0.3                        0    nvidia
[conda] cuda-libraries-dev        13.0.3                        0    nvidia
[conda] cuda-nvrtc                13.0.88              hb0a779f_0    nvidia
[conda] cuda-nvrtc-dev            13.0.88              hb0a779f_0    nvidia
[conda] cuda-nvtx                 13.0.85              hf384b4a_0    nvidia
[conda] cuda-opencl               13.0.85              hf384b4a_0    nvidia
[conda] cuda-opencl-dev           13.0.85              h3b4bcfc_0    nvidia
[conda] cudnn                     9.17.0.29            h45f0456_0  
[conda] libcublas                 13.1.1.3             hfa1462c_0    nvidia
[conda] libcublas-dev             13.1.1.3             hfa1462c_0    nvidia
[conda] libcudnn                  9.17.0.29            haeee033_0  
[conda] libcudnn-dev              9.17.0.29            h1e07607_0  
[conda] libcufft                  12.0.0.61            hf0c36ae_0    nvidia
[conda] libcufft-dev              12.0.0.61            hf0c36ae_0    nvidia
[conda] libcurand                 10.4.0.35            h50d9014_0    nvidia
[conda] libcurand-dev             10.4.0.35            h50d9014_0    nvidia
[conda] libcusolver               12.0.4.66            h3570045_0    nvidia
[conda] libcusolver-dev           12.0.4.66            h3570045_0    nvidia
[conda] libcusparse               12.6.3.3             h8296a5b_0    nvidia
[conda] libcusparse-dev           12.6.3.3             h8296a5b_0    nvidia
[conda] libnvjitlink              13.0.88              hb0a779f_0    nvidia
[conda] libnvjitlink-dev          13.0.88              hb0a779f_0    nvidia
[conda] nccl                      2.30.3.1             h6e9c2f8_0    conda-forge
[conda] nccl4py                   0.3.1                    pypi_0    pypi
[conda] numpy                     2.4.4                    pypi_0    pypi
[conda] optree                    0.19.0                   pypi_0    pypi
[conda] torch                     2.14.0a0+gitdd30fd7          pypi_0    pypi
[conda] triton                    3.6.0                    pypi_0    pypi
```

cc: @kwen2501 

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy @kapilsh

### PR DESCRIPTION
**Pull request by RohitRathore1**
`put_signal()/wait_signal()` validate the channel argument (since #191596) but pass `dst_rank/src_rank` from Python to the device kernels unchecked. The rank indexes the device array of signal pad pointers (`put_signal`) and the per-channel slot row (`wait_signal`), so an out-of-range value dereferences a wild pointer or lands one row past the signal pad but in the tensor data with the pad-first layout. I verified on 2x H200: `wait_signal(src_rank=world_size, channel=max)` was accepted and silently consumed the caller's own tensor element via the CAS(1 -> 0) protocol. The mistake is easy to hit in practice by passing a global rank where a subgroup-local rank is expected.

The fix adds `check_rank()` next to `check_channel()` and calls it host-side from both the CUDA and NCCL backends, so a bad rank raises before any kernel launch. An alternative was a single check at the pybind layer covering all backends at once; the per-backend helper mirrors how `check_channel()` is already wired, so it stays consistent with the existing structure. Also corrects the `check_channel()` negative-channel error message, which said "greater than 0" for a >= 0 check.

Tests also close the coverage gap from #191618: #191596 only tested `barrier()` on the CUDA backend, so this adds channel OOB tests for `put_signal()/ wait_signal()` (the demonstrated corruption vector), rank OOB tests for both backends, and a channel OOB test for the NCCL backend's `barrier()`.

Closes #191618

Test Plan:

```
python test/distributed/test_symmetric_memory.py -k "out_of_bounds"
python test/distributed/test_nccl.py -k "symmem"
```

### CODE REVIEW

**Review by Mr-Ashish — COMMENTED** [review]
## 🏴‍☠️ Torii — COMMENT
**Score:** 40/100 · **Confidence:** low

Torii: COMMENT

_Full review is the PR comment with marker `<!-- torii-review pr=191842`._

**Review by Mr-Ashish — COMMENTED** [review]
## 🏴‍☠️ Torii — COMMENT
**Score:** 40/100 · **Confidence:** low

Torii: COMMENT

_Full review is the PR comment with marker `<!-- torii-review pr=191842`._

**Review by Mr-Ashish — COMMENTED** [review]
## 🏴‍☠️ Torii — COMMENT
**Score:** 40/100 · **Confidence:** low

Torii: COMMENT

_Full review is the PR comment with marker `<!-- torii-review pr=191842`._

**Review by Mr-Ashish — COMMENTED** [review]
## 🏴‍☠️ Torii — COMMENT
**Score:** 40/100 · **Confidence:** low

Torii: COMMENT

_Full review is the PR comment with marker `<!-- torii-review pr=191842`._

**Review by Mr-Ashish — COMMENTED** [review]
## 🏴‍☠️ Torii — COMMENT
**Score:** 40/100 · **Confidence:** low

Torii: COMMENT

_Full review is the PR comment with marker `<!-- torii-review pr=191842`._

**Review by Mr-Ashish — COMMENTED** [review]
## 🏴‍☠️ Torii — COMMENT
**Score:** 40/100 · **Confidence:** low

Torii: COMMENT

_Full review is the PR comment with marker `<!-- torii-review pr=191842`._

**Review by Mr-Ashish — COMMENTED** [review]
## 🏴‍☠️ Torii — COMMENT
**Score:** 40/100 · **Confidence:** low

Torii: COMMENT

_Full review is the PR comment with marker `<!-- torii-review pr=191842`._

**Review by Mr-Ashish — COMMENTED** [review]
## 🏴‍☠️ Torii — COMMENT
**Score:** 40/100 · **Confidence:** low

Torii: COMMENT

_Full review is the PR comment with marker `<!-- torii-review pr=191842`._

**Review by Mr-Ashish — APPROVED** [review]
## 🏴‍☠️ Torii — APPROVE
**Score:** 88/100 · **Confidence:** high

Torii: APPROVE

_Full review is the PR comment with marker `<!-- torii-review pr=191842`._

**Comment by RohitRathore1** [comment]
hey @kapilsh please review this PR, thanks!

**Comment by kapilsh** [comment]
@claude can you check if the fix applies to all call sites that could have the out-of-bounds issue? also validate any perf implications/regressions possible from this? additionally, double check all ifdefs to ensure checks are enclosed in right ifdefs everywhere

**Comment by kapilsh** [comment]
@RohitRathore1 can you please fix the couple of things flagged above? they seem legit.

**Review by kapilsh — CHANGES_REQUESTED** [review]
- Address coverage `SymmetricMemory::get_signal_pad()` has the same issue
- Confirm whether pieces of `check_rank` logic need to be wrapped in appropriate macros

**Comment by RohitRathore1** [comment]
> @RohitRathore1 can you please fix the couple of things flagged above? they seem legit.

sure, looking into it

**Comment by RohitRathore1** [comment]
hi @kapilsh, i have addressed all the comments, thanks!

**Review by kapilsh — APPROVED** [review]
